mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala

```diff
@@ -129,17 +129,21 @@ object Bucketizer extends DefaultParamsReadable[Bucketizer] {
     if (feature == splits.last) {
       splits.length - 2
     } else {
-      val idx = ju.Arrays.binarySearch(splits, feature)
-      if (idx >= 0) {
-        idx
+      if (feature.isNaN) {
+        splits.length - 1
```

Comment:
index NaN value as ${splits.length -1}
''' all the other changes within this files, without comments, are just for code refactoring